flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/source/PulsarSourceReaderTestBase.java

```diff
@@ -0,0 +1,232 @@
+import static org.apache.flink.connector.pulsar.source.PulsarSourceOptions.PULSAR_SUBSCRIPTION_TYPE;
+import static org.apache.flink.connector.pulsar.source.reader.deserializer.PulsarDeserializationSchema.pulsarSchema;
+import static org.apache.flink.connector.pulsar.testutils.PulsarTestCommonUtils.createPartitionSplit;
+import static org.apache.flink.shaded.guava30.com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+abstract class PulsarSourceReaderTestBase extends PulsarTestSuiteBase {
+
+    protected static final int DEFAULT_POLLING_TIMEOUT_SECONDS = 120;
+
+    /** Expected the add or override config entries in inherited classes. */
+    protected abstract void customConfig(Configuration config);
+
+    @BeforeEach
+    void beforeEach(TestInfo testInfo) {
+        Random random = new Random(System.currentTimeMillis());
+        String topicName = uniqueTestName(testInfo);
+        operator().setupTopic(topicName, Schema.INT32, () -> random.nextInt(20));
+    }
+
+    @AfterEach
+    void afterEach(TestInfo testInfo) {
+        String topicName = uniqueTestName(testInfo);
+        operator().deleteTopic(topicName, true);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void assignZeroSplitsCreatesZeroSubscription(
+            boolean autoAcknowledgementEnabled, TestInfo testInfo) throws Exception {
+        PulsarSourceReaderBase<Integer> reader = sourceReader(autoAcknowledgementEnabled);
+        String topicName = uniqueTestName(testInfo);
+        reader.snapshotState(100L);
+        reader.notifyCheckpointComplete(100L);
+        // Verify the committed offsets.
+        reader.close();
+        for (int i = 0; i < PulsarRuntimeOperator.DEFAULT_PARTITIONS; i++) {
+            verifyNoSubscriptionCreated(TopicNameUtils.topicNameWithPartition(topicName, i));
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void assigningEmptySplits(boolean autoAcknowledgementEnabled, TestInfo testInfo)
+            throws Exception {
+        PulsarSourceReaderBase<Integer> reader = sourceReader(autoAcknowledgementEnabled);
+        String topicName = uniqueTestName(testInfo);
+        final PulsarPartitionSplit emptySplit =
+                createPartitionSplit(
+                        topicName, 0, Boundedness.CONTINUOUS_UNBOUNDED, MessageId.latest);
+
+        reader.addSplits(Collections.singletonList(emptySplit));
+
+        TestingReaderOutput<Integer> output = new TestingReaderOutput<>();
+        InputStatus status;
+        status = reader.pollNext(output);
+        // We sleep for 1 sec to avoid entering false negative case:
+        // we perform assertion when add splits task
+        // is dequeued but not executed.
+        sleepUninterruptibly(1, TimeUnit.SECONDS);
```

Comment:
Can you explain why we need this? This also seems like a source of flakiness.

Reply:
Oh, there is an edge case. Basically the fetcher could become idle upon entry because : 1. the AddSplitsTask is dequeued from taskQueue, 2. The dequeued splits is not assigned to `assignedSplits`.And if both `assignedSplits` is null and taskQueue is empty, the fetcher could become idle, and return NOTHING_AVAILABLE upon entry. This happened multiple times during local test.